geomesa-convert/geomesa-convert-common/src/main/scala/org/locationtech/geomesa/convert2/TypeInference.scala

```diff
@@ -0,0 +1,413 @@
+    *                    and still consider the column to be that type
+    * @return
+    */
+  def infer(data: Iterable[Iterable[Any]],
+            names: Iterable[String] = Seq.empty,
+            failureRate: Float = 0.1f): IndexedSeq[InferredType] = {
+    // list of inferred types for each column
+    val rawTypes = ArrayBuffer.empty[ListBuffer[InferredType]]
+    var i = 0
+    data.iterator.foreach { row =>
+      // skip empty rows or rows consisting of a single whitespace-only string
+      if (row.nonEmpty && (row.size > 1 || row.headOption.collect { case s: String if s.trim.isEmpty => s }.isEmpty)) {
+        i = 0
+        row.foreach { col =>
+          if (i == rawTypes.length) {
+            // grow our column array if needed, we don't assume all rows have the same number of columns
+            rawTypes += ListBuffer.empty[InferredType]
+          }
+          // add the inferred type for this column value - it won't have a name at this point
+          rawTypes(i) += InferredType.infer(col)
+          i += 1
+        }
+      }
+    }
+
+    // track the names we use for each column to ensure no duplicates
+    val uniqueNames = scala.collection.mutable.HashSet.empty[String]
+    // names that were provided
+    val nameIterator = names.iterator
+
+    val types = rawTypes.map { raw =>
+
+      // merge the types for this column from each row
+      val typ = merge(raw, failureRate)
+
+      // determine the column name
+      var base: String = null
+      var name: String = null
+      var i = 0
+
+      if (nameIterator.hasNext) {
+        // if the name was passed in, start with that
+        base = nameIterator.next
+        name = base
+      } else {
+        // use the type with an _i, e.g. int_0, string_1
+        base = typ.typed.toString.toLowerCase(Locale.US)
+        name = s"${base}_0"
+        i += 1
+      }
+
+      while (FeatureUtils.ReservedWords.contains(name.toUpperCase(Locale.US)) || !uniqueNames.add(name)) {
+        name = s"${base}_$i"
+        i += 1
+      }
+
+      // set the name for the column
+      typ.copy(name = name)
+    }
+
+    // if there is no geometry field, see if we can derive one
+    if (types.lengthCompare(1) > 0 && !types.map(_.typed).exists(geometries.contains)) {
+      var name = "geom"
+      var i = 0
+      while (!uniqueNames.add(name)) {
+        name = s"geom_$i"
+        i += 1
+      }
+
+      var lat, lon: String = null
+
+      // if we have lat/lon columns, use those
+      types.exists { t =>
+        val name = t.name.toLowerCase(Locale.US)
+        if (latitudeNames.contains(name)) {
+          lat = t.name
+        } else if (longitudeNames.contains(name)) {
+          lon = t.name
+        }
+        // stop iterating if we find them both
+        lat != null && lon != null
+      }
+
+      if (lat == null || lon == null) {
+        // as a fallback, check for 2 consecutive floats or doubles - assume longitude first
```

Comment:
Also, around these assumptions, is there a check for the size of the values?  A -100 in one column might provide evidence that something is not a latitude...

Reply:
yeah, that's a good idea - i don't think there are checks like that yet

Reply:
opened a ticket https://geomesa.atlassian.net/browse/GEOMESA-2330